### ISSUE TITLE
Collection was modified (ObjectGraphScanner.ScanProperties)

### ISSUE THREAD

**Issue by paulem** [NLog/NLog#360]
We used NLog 2.1 in a multithreaded environment on .NET Framework 4 
and got this error:

```
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at System.Collections.Generic.List`1.Enumerator.MoveNext()
   at NLog.Internal.EnumerableHelpers.<OfType>d__0`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at NLog.Internal.EnumerableHelpers.ToList[T](IEnumerable`1 enumerable)
   at NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, Dictionary`2 visitedObjects)
   at NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, Dictionary`2 visitedObjects)
   at NLog.Internal.ObjectGraphScanner.FindReachableObjects[T](Object[] rootObjects)
   at NLog.Internal.TargetWithFilterChain.PrecalculateStackTraceUsage()
   at NLog.LogFactory.GetTargetsByLevelForLogger(String name, IList`1 rules, TargetWithFilterChain[] targetsByLevel, TargetWithFilterChain[] lastTargetsByLevel)
   at NLog.LogFactory.GetTargetsByLevelForLogger(String name, IList`1 rules, TargetWithFilterChain[] targetsByLevel, TargetWithFilterChain[] lastTargetsByLevel)
   at NLog.LogFactory.GetConfigurationForLogger(String name, LoggingConfiguration configuration)
   at NLog.LogFactory.GetLogger(LoggerCacheKey cacheKey)
   at NLog.LogFactory.GetLogger(String name)
   at NLog.LogManager.GetCurrentClassLogger()

```

![screenshot 2014-06-10 11 06 32](https://cloud.githubusercontent.com/assets/2874236/3226520/56e2c0f8-f06e-11e3-84ce-7595801b904d.png)

Here is the config:

![image](https://cloud.githubusercontent.com/assets/2874236/3226575/8c044bf2-f06f-11e3-8d3d-cec2027b7e83.png)

**Comment by paulem** [comment]
Hi @Xharze, 
the error has reappeared in 3.2.0.0

```System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
 в System.Collections.Generic.List`1.Enumerator.MoveNextRare()
 в System.Collections.Generic.List`1.Enumerator.MoveNext()
 в NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, Dictionary`2 visitedObjects)
 в NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, Dictionary`2 visitedObjects)
 в NLog.Internal.ObjectGraphScanner.FindReachableObjects[T](Object[] rootObjects)
 в NLog.Internal.TargetWithFilterChain.PrecalculateStackTraceUsage()
 в NLog.LogFactory.GetTargetsByLevelForLogger(String name, IList`1 rules, TargetWithFilterChain[] targetsByLevel, TargetWithFilterChain[] lastTargetsByLevel)
 в NLog.LogFactory.GetConfigurationForLogger(String name, LoggingConfiguration configuration)
 в NLog.LogFactory.GetLogger(LoggerCacheKey cacheKey)
 в NLog.LogFactory.GetLogger(String name)
 в NLog.LogManager.GetCurrentClassLogger()```

**Comment by 304NotModified** [comment]
This is probably the same as #500 with fix at#704

**Comment by 304NotModified** [comment]
This is the same issue. So almost done

**Comment by 304NotModified** [comment]
Will be fixed in the 4.0 release.

**Comment by 304NotModified** [comment]
PR #704 merged, closing this
